tests/integration/test_database.py

```diff
@@ -241,6 +255,84 @@ async def test_relation_creation(ops_test: OpsTest):
 @pytest.mark.order(6)
 @pytest.mark.abort_on_fail
 @pytest.mark.database_tests
+async def test_ready_only_endpoints(ops_test: OpsTest):
+    """Check read-only-endpoints are correctly updated."""
+    relation_data = await get_relation_data(
+        ops_test=ops_test, application_name=DATABASE_APP_NAME, relation_name=DB_RELATION_NAME
+    )
+    assert len(relation_data) == 1
+    read_only_endpoints = get_read_only_endpoint_hostnames(relation_data)
+    # check correct number of read-only-endpoint is correct
+    assert len(ops_test.model.applications[DATABASE_APP_NAME].units) - 1 == len(
+        read_only_endpoints
+    )
+    app_hostnames = await get_unit_hostname(ops_test=ops_test, app_name=DATABASE_APP_NAME)
+    # check that endpoints are the one of the application
+    for r_endpoint in read_only_endpoints:
+        assert r_endpoint in app_hostnames
+
+    # increase the number of units
+    async with ops_test.fast_forward():
+        await scale_application(ops_test, DATABASE_APP_NAME, 4)
+    # check update for read-only-endpoints
+    relation_data = await get_relation_data(
+        ops_test=ops_test, application_name=DATABASE_APP_NAME, relation_name=DB_RELATION_NAME
+    )
+    read_only_endpoints = get_read_only_endpoint_hostnames(relation_data)
+    # check that the number of read-only-endpoints is correct
+    assert len(ops_test.model.applications[DATABASE_APP_NAME].units) - 1 == len(
+        read_only_endpoints
+    )
+    app_hostnames = await get_unit_hostname(ops_test=ops_test, app_name=DATABASE_APP_NAME)
+    # check that endpoints are the one of the application
+    for r_endpoint in read_only_endpoints:
+        assert r_endpoint in app_hostnames
+
+    # decrease the number of units
+    async with ops_test.fast_forward():
+        await scale_application(ops_test, DATABASE_APP_NAME, 2)
+
+    # wait for the update of the endpoints
+    time.sleep(2 * 60)
```

Comment:
we generally prefer not to sleep in tests as this is not deterministic. you can instead follow a retry pattern (with tenacity) where you retry if you attempt to get something and it is not present, or use `ops_test.block_until` until a condition is met

[here](https://github.com/canonical/mysql-k8s-operator/blob/2c0c7c5647f3fdd5d7a6e81f5a006ec1f2f9283b/tests/integration/test_charm.py#L110-L125) is an example of the retry pattern